### PR TITLE
Force include lib/ files with full path

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
There are two ways to include files under `src/lib`, `#include <lib/core/xxx.h>` or `#include <core/xxx.h>`. It turns out that mixing these two include may confuse `#pragma once` and causing redefinition error.

#### Change overview
Disallow using #include <core/xxx.h>. force use the full path with 'lib/'.

#### Testing
Verified using unit-tests

### CODE REVIEW

**Comment by kghost** [comment]
I use following command to convert includes:

```bash
for i in support core asn1 mdns shell; do git sed "#include \"$i/\\(.*\\)\"\$" "#include <lib/$i/\\1>" ; done
for i in support core asn1 mdns shell; do git sed "#include <$i/\\(.*\\)>\$" "#include <lib/$i/\\1>" ; done
```

File: src/lib/shell/streamer_zephyr.cpp [hunk 1]
```diff
@@ -24,8 +24,8 @@
 
 #include <cassert>
 
-#include <shell/shell.h>
-#include <shell/shell_uart.h>
+#include <lib/shell/shell.h>
+#include <lib/shell/shell_uart.h>
```
Comment (bzbarsky-apple) [inline src/lib/shell/streamer_zephyr.cpp]: Neither of those files is in our tree, so this change is wrong.

**Comment by woody-apple** [comment]
Note, @kghost - this has real build failures.

File: src/lib/shell/streamer_zephyr.cpp (hunk 1 above)
Comment (kghost) [reply src/lib/shell/streamer_zephyr.cpp]: Fixed

File: src/lib/shell/MainLoopZephyr.cpp [hunk 2]
```diff
@@ -15,7 +15,7 @@
  *    limitations under the License.
  */
 #include <init.h>
-#include <shell/shell.h>
+#include <lib/shell/shell.h>
```
[the remark this answers is not in the document]
Comment (kghost) [reply src/lib/shell/MainLoopZephyr.cpp]: Fixed

File: src/tools/chip-cert/chip-cert.h [hunk 3]
```diff
@@ -50,19 +50,19 @@
 #include <openssl/x509v3.h>
 
 #include <CHIPVersion.h>
-#include <asn1/ASN1.h>
 #include <asn1/ASN1OID.h>
-#include <core/CHIPConfig.h>
-#include <core/CHIPCore.h>
-#include <core/CHIPSafeCasts.h>
 #include <credentials/CHIPCert.h>
```
[the remark this answers is not in the document]
Comment (kghost) [reply src/tools/chip-cert/chip-cert.h:53]: Fixed

File: examples/platform/nrfconnect/util/DFUTrigger.cpp [hunk 4]
```diff
@@ -21,7 +21,7 @@
 #endif
 
 #include <drivers/gpio.h>
-#include <shell/shell.h>
+#include <lib/shell/shell.h>
```
[the remark this answers is not in the document]
Comment (kghost) [reply examples/platform/nrfconnect/util/DFUTrigger.cpp]: Fixed

**Comment by bzbarsky-apple** [comment]
> I use following command to convert includes:

Please:

1. Document in the commit message that these are the commands that were run.
2. Document in the commit message that the git-extras package is needed to run them.
3. Document in the commit message what manual fixups were then performed after running the commands.

**Review by bzbarsky-apple — COMMENTED** [review]
So comparing this to the output of running the commands claimed and doing the fixups that are needed to fix the build:

1. Why the blank line additions to `src/app/zap-templates/templates/app/af-structs.zapt`?
2. Why the removal of the `ASN1OID.h` include from `src/credentials/tests/CHIPCert_test_vectors.h`?
4. Why the removal of the `ASN1OID.h` include from `src/lib/asn1/tests/TestASN1.cpp`?
5. Why the removal of the `ASN1OID.h` include from `src/tools/chip-cert/chip-cert.h`?

**Comment by kghost** [comment]
> 3. Document in the commit message what manual fixups were then performed after running the commands.

I'm already unable to trace manual fixes in the PR. It is around ten or twenty manual fixes.

I can make sure that only includes are changes, and their order due to restyle.

**Comment by kghost** [comment]
> So comparing this to the output of running the commands claimed and doing the fixups that are needed to fix the build:
> 
> 1. Why the blank line additions to `src/app/zap-templates/templates/app/af-structs.zapt`?

Keep restyle from change its order. 

> 2. Why the removal of the `ASN1OID.h` include from `src/credentials/tests/CHIPCert_test_vectors.h`?
> 3. Why the removal of the `ASN1OID.h` include from `src/lib/asn1/tests/TestASN1.cpp`?
> 4. Why the removal of the `ASN1OID.h` include from `src/tools/chip-cert/chip-cert.h`?

`ASN1OID.h` is included by `src/lib/asn1/ASN1.h` inside `chip::ASN1` namespace 

```
namespace chip {
namespace ASN1 {

#ifndef DOXYGEN_SHOULD_SKIP_THIS
#include <asn1/ASN1OID.h>
#endif

```

Above includes doesn't put the file in the namespace, so causing compile error, it is probably a bug hidden behind the include order.

**Comment by bzbarsky-apple** [comment]
> I'm already unable to trace manual fixes in the PR. It is around ten or twenty manual fixes.

There are manual fixes in 30 files or so, most of them in just a few categories.  For future, to make this sort of thing reviewable, I would recomment one changeset with the script-generated mass-change, documented with what the scripts involved were (so reviewers can easily check that nothing else is in that changeset; in this case that scripted mass-change would be the sed scripts, restyled-io, and ZAP codegen) and one changeset with the manual changes that actually need to be manually reviewed....

> I can make sure that only includes are changes, and their order due to restyle.

Right, given the scripts I manually verified things; that's where https://github.com/project-chip/connectedhomeip/pull/9387#pullrequestreview-744049185 comes from.  :)

**Comment by bzbarsky-apple** [comment]
> ASN1OID.h is included by src/lib/asn1/ASN1.h inside chip::ASN1 namespace

Oh, wow.  That is ... not sane.  Please file a followup issue to fix that and just put the right namespacing info in the generated file instead of doing that?

**Review by bzbarsky-apple — APPROVED** [review]
Approved on condition that the commit message is updated as I requested:

1. The scripts used
2. The git-extras package needed
3. General summary of manual changes (undo the changes for the SDK shell stuff, modify gn files to actually do the forcing of full path, stop including the IOD header outside of the namespace bits, that seems like about it?)

**Comment by bzbarsky-apple** [comment]
And thank you for doing this!

**Comment by kghost** [comment]
> There are manual fixes in 30 files or so, most of them in just a few categories. For future, to make this sort of thing reviewable, I would recomment one changeset with the script-generated mass-change, documented with what the scripts involved were (so reviewers can easily check that nothing else is in that changeset; in this case that scripted mass-change would be the sed scripts, restyled-io, and ZAP codegen) and one changeset with the manual changes that actually need to be manually reviewed....

I'll split them into 2 patches, but the script generated patch will not compile, so it is unable to separate them into 2 PRs

> Oh, wow. That is ... not sane. Please file a followup issue to fix that and just put the right namespacing info in the generated file instead of doing that?

Not problem

**Comment by mspang** [comment]
The reason this didn't happen in the last round of cleanups is that lib/ itself is inconsistent. What criteria defines whether something goes in lib/ or not? I can't think of one.

So there are two options
- Add lib to includes. This still seems inconsistent.
- Move the contents of lib/ up one level. In this case we don't need to change the vast majority of #include lines.

**Comment by kghost** [comment]
This goal of this PR is to disambiguate the include path of `lib/core/xxx.h` and `core/xxx.h`. It is not going to do any cleanups.

**Review by andy31415 — APPROVED** [review]
Approve assuming CI pases and changes are only include changes. I am unable to properly review because Github UI uses 100% CPU and fails to easily show me the deltas.

**Comment by andy31415** [comment]
@kghost - the mbedos failure may be real.

**Review by yufengwangca — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by mspang** [comment]
> Approve assuming CI pases and changes are only include changes. I am unable to properly review because Github UI uses 100% CPU and fails to easily show me the deltas.

You can download the patch..

**Comment by bzbarsky-apple** [comment]
Indeed.  And given that @kghost split this up into multiple changesets, can verify that the script-generated changeset matches the script output and review the small manual bits....  That's what I ended up doing.

**Pull request merged**